### ISSUE TITLE
Early EOF Exception in Jetty Http Client. Connection Idle Timeout 180 Sec

### ISSUE THREAD

**Issue by gupta-sumit** [eclipse/jetty.project#4125]
Hi 

We are getting early eof exception in jetty http client. Not sure what is causing this.  Can you please help us.

```
 Error occured; nested exception is: 
	java.util.concurrent.ExecutionException: java.io.EOFException: HttpConnectionOverHTTP@46742eef(l:/10.56.4.47:43448 <-> r:g8prodr4xapi.navitaire.com/170.248.209.2:443,closed=false)=>HttpChannelOverHTTP@70312e5e(exchange=HttpExchange@1f4407b7 req=TERMINATED/null@null res=PENDING/null@null)[send=HttpSenderOverHTTP@39b251cb(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@602f319{s=START}],recv=HttpReceiverOverHTTP@61e6b660(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]<-DecryptedEndPoint@185c27dc{g8prodr4xapi.navitaire.com/170.258.209.2:443<->/10.56.4.47:43448,OPEN,fill=-,flush=-,to=30283/180000}->HttpConnectionOverHTTP@46742eef(l:/10.56.4.47:43448 <-> r:g8prodr4xapi.navitaire.com/170.258.209.2:443,closed=false)=>HttpChannelOverHTTP@70312e5e(exchange=HttpExchange@1f4407b7 req=TERMINATED/null@null res=PENDING/null@null)[send=HttpSenderOverHTTP@39b251cb(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@602f319{s=START}],recv=HttpReceiverOverHTTP@61e6b660(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]->SocketChannelEndPoint@5613fb28{g8prodr4xapi.navitaire.com/170.258.209.2:443<->/10.56.4.47:43448,ISHUT,fill=-,flush=-,to=132/180000}{io=0/0,kio=0,kro=1}->SslConnection@33cb0115{NEED_WRAP,eio=-1/-1,di=-1}=>HttpConnectionOverHTTP@46742eef(l:/10.56.4.47:43448 <-> r:g8prodr4xapi.navitaire.com/170.258.209.2:443,closed=false)=>HttpChannelOverHTTP@70312e5e(exchange=HttpExchange@1f4407b7 req=TERMINATED/null@null res=PENDING/null@null)[send=HttpSenderOverHTTP@39b251cb(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@602f319{s=START}],recv=HttpReceiverOverHTTP@61e6b660(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]
```

**Comment by joakime** [comment]
The stacktrace will tell you what the cause is.

Early EOF just means the HTTP Protocol terminated before it was expected to.
From a server point of view, this is actually very common from mobile clients.
From a client point of view, this is common with bad/incomplete HTTP responses.

**Comment by gupta-sumit** [comment]
Is there any option in Jetty Http Client to test connection before borrow from connection pool?

**Comment by sbordet** [comment]
> Is there any option in Jetty Http Client to test connection before borrow from connection pool?

@gupta-sumit that is not the problem, we already do that.
Please post the full stack trace.

**Comment by gupta-sumit** [comment]
aused by: java.io.EOFException: HttpConnectionOverHTTP@7f09dd97(l:/10.56.13.205:60548 <-> r:sgprodr4xapi.navitaire
.com/170.248.238.62:443,closed=false)=>HttpChannelOverHTTP@177a12f9(exchange=HttpExchange@5dd838bc req=PENDING/null
@null res=PENDING/null@null)[send=HttpSenderOverHTTP@49ef7dc7(req=HEADERS,snd=SENDING,failure=null)[HttpGenerator@3
4eea21e{s=COMPLETING}],recv=HttpReceiverOverHTTP@53a3471(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]<-Dec
ryptedEndPoint@421013f5{sgprodr4xapi.navitaire.com/170.248.238.62:443<->/10.56.13.205:60548,OPEN,fill=-,flush=P,to=
588/60000}->HttpConnectionOverHTTP@7f09dd97(l:/10.56.13.205:60548 <-> r:sgprodr4xapi.navitaire.com/170.248.238.62:4
43,closed=false)=>HttpChannelOverHTTP@177a12f9(exchange=HttpExchange@5dd838bc req=PENDING/null@null res=PENDING/nul
l@null)[send=HttpSenderOverHTTP@49ef7dc7(req=HEADERS,snd=SENDING,failure=null)[HttpGenerator@34eea21e{s=COMPLETING}
],recv=HttpReceiverOverHTTP@53a3471(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]->SocketChannelEndPoint@3a
bc9ef7{sgprodr4xapi.navitaire.com/170.248.238.62:443<->/10.56.13.205:60548,ISHUT,fill=-,flush=-,to=580/60000}{io=0/
0,kio=0,kro=1}->SslConnection@70a35f35{NEED_WRAP,eio=-1/-1,di=-1}=>HttpConnectionOverHTTP@7f09dd97(l:/10.56.13.205:
60548 <-> r:sgprodr4xapi.navitaire.com/170.248.238.62:443,closed=false)=>HttpChannelOverHTTP@177a12f9(exchange=Http
Exchange@5dd838bc req=PENDING/null@null res=PENDING/null@null)[send=HttpSenderOverHTTP@49ef7dc7(req=HEADERS,snd=SEN
DING,failure=null)[HttpGenerator@34eea21e{s=COMPLETING}],recv=HttpReceiverOverHTTP@53a3471(rsp=IDLE,failure=null)[H
ttpParser{s=CLOSED,0 of -1}]]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.earlyEOF(HttpReceiverOverHTTP.java:321) ~[jetty-clien
t-9.4.8.v20171121.jar!/:9.4.8.v20171121]
        at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1557) ~[jetty-http-9.4.11.v20180605.jar!/:9.
4.11.v20180605]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.shutdown(HttpReceiverOverHTTP.java:197) ~[jetty-clien
t-9.4.8.v20171121.jar!/:9.4.8.v20171121]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:144) ~[jetty-client
-9.4.8.v20171121.jar!/:9.4.8.v20171121]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:70) ~[jetty-client-
9.4.8.v20171121.jar!/:9.4.8.v20171121]
        at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:137) ~[jetty-client-9
.4.8.v20171121.jar!/:9.4.8.v20171121]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:155) ~[jetty
-client-9.4.8.v20171121.jar!/:9.4.8.v20171121]
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281) ~[jetty-io-9
.4.11.v20180605.jar!/:9.4.11.v20180605]
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102) ~[jetty-io-9.4.11.v20180605.jar!/:9.4.
11.v20180605]
        at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:291) ~[jetty-io-9.4.11.v20180605.ja
r!/:9.4.11.v20180605]

**Comment by gupta-sumit** [comment]
@sbordet I have attached stack trace.

**Comment by gupta-sumit** [comment]
@sbordet How you classifying below types using exception. As if we need to retry on some transaction based on exception, exception should be clear

Connect Timeout : I am getting TimeoutException on this.
Socket Read Timeout: Server might have processed request but client got timed out. How do I identify this.
Connection Reset: Is EOF exception always means Connection Reset. And is valid for retry always

**Comment by joakime** [comment]
jetty-client-9.4.8.v20171121.jar
jetty-io-9.4.11.v20180605.jar

You have a mismatch in Jetty versions, fix that first, as that's really bad.

Next, please update to a more recent version of Jetty and try again.
Consider using 9.4.20.v20190813

**Comment by gupta-sumit** [comment]
@joakime Thanks for noticing version mismatch. But will it really cause issue??

**Comment by sbordet** [comment]
The log/stack shows that the client sent the request headers, and the server closed the connection.
You have to look at the server logs to know why the server closed the connection.

**Comment by sirianni** [comment]
We are running into a similar issue.  Like others, we are running Jetty on both the client and the server.  On the client side we are using Jetty's `AsyncProxyServlet`.  This is running in the context of [apache/druid](https://github.com/apache/druid).

```
java.io.EOFException: HttpConnectionOverHTTP@55bbf074::SocketChannelEndPoint@2363c4e6{druid-prod-broker-2.druid-prod-broker.druid-prod.svc.cluster.local/100.96.85.79:8082<->/100.96.171.162:39462,ISHUT,fill=-,flush=-,to=6493/300000}{io=0/0,kio=0,kro=1}->HttpConnectionOverHTTP@55bbf074(l:/100.96.171.162:39462 <-> r:druid-prod-broker-2.druid-prod-broker.druid-prod.svc.cluster.local/100.96.85.79:8082,closed=false)=>HttpChannelOverHTTP@d4e73c7(exchange=HttpExchange@8f3e0d5 req=TERMINATED/null@null res=PENDING/null@null)[send=HttpSenderOverHTTP@5efdeaae(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@63844163{s=START}],recv=HttpReceiverOverHTTP@1d2ac69f(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.earlyEOF(HttpReceiverOverHTTP.java:338)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1551)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.shutdown(HttpReceiverOverHTTP.java:209)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:147)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:73)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:133)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:155)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
	at java.lang.Thread.run(Thread.java:748)
```

We have started seeing this issue more frequently under load.  Trying to parse through the `toString()` of the `HttpConnectionOverHTTP` object, it seems like the server closed the connection unexpectedly (?), however we see no errors in the Jetty server logs.
```
  HttpConnectionOverHTTP@55bbf074::SocketChannelEndPoint@2363c4e6
    {
      druid-prod-broker-2.druid-prod-broker.druid-prod.svc.cluster.local/100.96.85.79:8082<->/100.96.171.162:39462,
      ISHUT,
      fill=-,
      flush=-,
      to=6493/300000
    }
    {
      io=0/0,
      kio=0,
      kro=1
    }
->
  HttpConnectionOverHTTP@55bbf074(
    l:/100.96.171.162:39462 <-> r:druid-prod-broker-2.druid-prod-broker.druid-prod.svc.cluster.local/100.96.85.79:8082,
    closed=false
  )
  =>
  HttpChannelOverHTTP@d4e73c7(exchange=HttpExchange@8f3e0d5 req=TERMINATED/null@null res=PENDING/null@null)
    [
      send=HttpSenderOverHTTP@5efdeaae(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@63844163{s=START}],
      recv=HttpReceiverOverHTTP@1d2ac69f(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]
    ]
```

**Comment by bhuvangu** [comment]
@gupta-sumit how did you solved it. I think it the problem of `sgprodr4xapi.navitaire.com` . Did you find any config/solution to solve it
